File: rules.ts
import * as fs from 'fs';
import * as yaml from 'js-yaml';

const ruleFile : string = process.env.RULE_FILE || 'rules.yml'

export interface ruleInterface {
    subreddit : string,
    type: string,
    titles: string[],
    bodies: string[],
    comments?: string[],
    every: number
}


function getRuleYaml(): any {
    try {
        return yaml.safeLoad(fs.readFileSync(ruleFile).toString())
    } catch (err) {
        console.log(`[FATAL] Error while reading rule file: ${err}`)
        process.exit(1)
    }
}

export async function parseRules(callback : any) {
    var rules = getRuleYaml();
    for (var rule in rules['rules']) {
        var IRule : ruleInterface = {
            type: rules['rules'][rule]['type'],
            subreddit: rules['rules'][rule]['subreddit'],
            titles: rules['rules'][rule]['titles'],
            bodies: rules['rules'][rule]['bodies'],
            comments: rules['rules'][rule]['comments'],
            every: rules['rules'][rule]['every']
        }
        await callback(IRule);
    }
    return true;
}

export function validateRules() : boolean {
    var rules = getRuleYaml();
    for (var rule in rules['rules']) {
        if (!ruleValid(rules['rules'][rule])) {
            return false;
        }
    }
    return true;
}

function ruleValid(rule: any) {
    if (rule == null) {
        return false
    } else {
        if (!('subreddit' in rule) || !('titles' in rule) || !('bodies' in rule) || !('every' in rule) || !('type' in rule)) {
            return false
        }

        if (rule['subreddit'] === null || rule['titles'] === null || rule['bodies'] === null || rule['every'] === null || rule['type'] === null) {
            return false
        }

        if (rule['type'] === 'text' || rule['type'] === 'link') {
            return true
        }
    }
}

File: autopost/rules.py
import yaml

class Rule:
    """Rule object."""
    def __init__(self, subreddit, kind, titles, bodies, comments, flair, every):
        """Initializes a Rule class."""
        self.subreddit = subreddit
        self.kind = kind
        self.titles = titles
        self.bodies = bodies
        self.comments = comments
        self.flair = flair
        self.every = every

def parseRules(rules, callback):
    """
    Parses rules into interfaces and provides them as arg to callback.

    Parameters:
    rules (dict): Rules YAML dict
    callback (func): Callback to call with rule interfaces
    """
    for rule in rules['rules']:
        IRule = Rule(
            rules['rules'][rule]['subreddit'],
            rules['rules'][rule]['kind'],
            rules['rules'][rule]['titles'],
            rules['rules'][rule]['bodies'],
            rules['rules'][rule]['comments'],
            rules['rules'][rule]['flair'],
            rules['rules'][rule]['every']
        )
        callback(IRule)

def validateRules(rules):
    """
    Valiates the YAML rules file.

    Parameters:
    rules (dict): Rules YAML dict

    Returns:
    bool: False if invalid rules are present, otherwise True.
    """
    for rule in rules['rules']:
        if not ruleValid(rules['rules'][rule]):
            return False
    return True

def ruleValid(rule):
    """
    Checks whether or not a rule has all required properties.

    Parameters:
    rule (dict): A single rule

    Returns:
    bool: False if rule does not meet requirements, otherwise True
    """
    if rule == None:
        return False
    if not ('subreddit' in rule) or not ('titles' in rule) or not ('bodies' in rule) or not ('every' in rule) or not ('kind' in rule):
        return False
    if (rule['subreddit'] == None) or (rule['titles'] == None) or (rule['bodies'] == None) or (rule['every'] == None) or (rule['kind'] == None):
        return False

    if (rule['kind'] == 'text' or rule['kind'] == 'link'):
        return True

def getYAML(file):
    """
    Returns loaded YAML from a file.

    Parameters:
    file (string): Path to the YAML file

    Returns:
    dict: The loaded YAML
    """
    try:
        with open(file, 'r') as stream:
            try:
                return yaml.safe_load(stream)
            except yaml.YAMLError as err:
                exit(f'[FATAL] Unexpected YAML error: {err}')
    except FileNotFoundError:
        exit(f'[FATAL] File {file} could not be found.')